### PR TITLE
Add additional `HashEngine` types

### PR DESCRIPTION
**Pull request by tcharding**
Currently we are using a type alias for a few of the hash engines.
    
Type alias' allow for potential mixing of types, a struct can better serve our users with not much additional complexity or maintenance burden.
 
- As preparation add const constructors where possible to hash engines (excl. `HmacEngine`).
- Add a `sha256d::HashEngine` struct
- Add a `hash160::HashEngine` struct

If this goes in we can improve the `sha256t::HashEngine` in a similar manner.

### CODE REVIEW

File: hashes/src/sha256d.rs [hunk 1]
```diff
@@ -13,10 +13,30 @@ crate::internal_macros::hash_type! {
...
+    /// Creates a new SHA256d hash engine.
+    pub const fn new() -> Self { Self(sha256::HashEngine::new()) }
+}
+
+impl Default for HashEngine {
+    fn default() -> Self { Self::new() }
+}
+
+impl crate::HashEngine for HashEngine {
+    type MidState = sha256::Midstate;
```
Comment (Kixunil) [inline hashes/src/sha256d.rs:30]: Should we have a new type here as well?

**Review by Kixunil — APPROVED** [review]
ACK 51010777bf4b4e738348855d5e637ea8f5571570

File: hashes/src/sha256d.rs (hunk 1 above)
Comment (tcharding) [reply hashes/src/sha256d.rs:30]: I thought that all the midstate stuff is a bit up in the air still so I just put it like this for now. Happy to take direction on it though, @apoelstra?

Comment (Kixunil) [reply hashes/src/sha256d.rs:30]: Yeah, that's why I'm asking and not claiming it should be that way. :)

Comment (tcharding) [reply hashes/src/sha256d.rs:30]: Cool, thanks man. Thanks for your reviews today as well!

Comment (apoelstra) [reply hashes/src/sha256d.rs:30]: I could really go either way on this, conceptually. But I think that if we're gonna separate the engine types we ought to also separate the midstates.

Comment (apoelstra) [reply hashes/src/sha256d.rs:30]: But I'm also fine leaving it as-is for now. We can address it when we do https://github.com/rust-bitcoin/rust-bitcoin/issues/2918 (another "easy", or at least localized, `hashes` API change).

Comment (Kixunil) [reply hashes/src/sha256d.rs:30]: Now I'm thinking having midstate on sha256t is really strange. Not strictly wrong but if you're using it maybe you meant to use sh256 midstate or you should've used our trait. So I suspect not all hashes should even support it.

Comment (apoelstra) [reply hashes/src/sha256d.rs:30]: Yes, agreed. I think that probably only sha256 should. Easy to add to others later, even post-1.0.

But I also think this is a question for #2918, not this PR.

**Review by apoelstra — APPROVED** [review]
ACK 51010777bf4b4e738348855d5e637ea8f5571570 but will hold off on merging until tcharding indicates whether he wants to address the midstate thing

**Comment by tcharding** [comment]
I'd prefer to leave it till #2918 please.

**Pull request merged**